nano/node/election.cpp

```diff
@@ -48,24 +51,168 @@ void nano::election::confirm_once (nano::election_status_type type_a)
+		}
+		if (base_latency () * 10 < std::chrono::steady_clock::now () - last_vote)
+		{
+			last_vote = std::chrono::steady_clock::now ();
+			node.block_processor.generator.add (status.winner->hash ());
+		}
+	}
+}
+
+void nano::election::transition_active ()
+{
+	if (!state_change (nano::election::state_t::idle, nano::election::state_t::active))
+	{
+		if (base_latency () * 5 < std::chrono::steady_clock::now () - last_block)
+		{
+			last_block = std::chrono::steady_clock::now ();
+			node.network.flood_block (status.winner);
+		}
+	}
+}
+
+void nano::election::transition_idle ()
+{
+	auto state_l = state_m.load ();
+	if (state_l != nano::election::state_t::confirmed)
+	{
+		state_change (state_l, nano::election::state_t::idle);
 	}
 }
 
+bool nano::election::idle () const
+{
+	return state_m == nano::election::state_t::idle;
+}
+
 bool nano::election::confirmed ()
 {
-	return confirmed_m;
+	return state_m == nano::election::state_t::confirmed;
+}
+
+void nano::election::activate_dependencies ()
+{
+	auto transaction = node.store.tx_begin_read ();
+	bool escalated_l (false);
+	std::shared_ptr<nano::block> previous_l;
+	auto previous_hash_l (status.winner->previous ());
+	if (!previous_hash_l.is_zero () && node.active.blocks.find (previous_hash_l) == node.active.blocks.end ())
+	{
+		previous_l = node.store.block_get (transaction, previous_hash_l);
+		if (previous_l != nullptr && !node.block_confirmed_or_being_confirmed (transaction, previous_hash_l))
+		{
+			auto election = node.active.insert_impl (previous_l, true);
+			if (election.second)
+			{
+				election.first->transition_active ();
+				escalated_l = true;
+			}
+		}
+	}
+	/* If previous block not existing/not commited yet, block_source can cause segfault for state blocks
+				So source check can be done only if previous != nullptr or previous is 0 (open account) */
+	if (previous_hash_l.is_zero () || previous_l != nullptr)
+	{
+		auto source_hash_l (node.ledger.block_source (transaction, *status.winner));
+		if (!source_hash_l.is_zero () && source_hash_l != previous_hash_l && node.active.blocks.find (source_hash_l) == node.active.blocks.end ())
+		{
+			auto source_l (node.store.block_get (transaction, source_hash_l));
+			if (source_l != nullptr && !node.block_confirmed_or_being_confirmed (transaction, source_hash_l))
+			{
+				auto election = node.active.insert_impl (source_l, true);
+				if (election.second)
+				{
+					election.first->transition_active ();
+					escalated_l = true;
+				}
+			}
+		}
+	}
+	if (escalated_l)
+	{
+		update_dependent ();
+	}
+}
+
+void nano::election::broadcast_block ()
+{
+	if (base_latency () * 5 < std::chrono::steady_clock::now () - last_block)
+	{
+		if (!node.active.solicitor.broadcast (*this))
+		{
+			last_block = std::chrono::steady_clock::now ();
+		}
+	}
+}
+
+bool nano::election::transition_time ()
+{
+	bool result = false;
+	if (std::chrono::minutes (5) < std::chrono::steady_clock::now () - election_start)
```

Comment:
Seems like elections themselves should have an `expired` state_t, and assert that `transition_time` is not called once they are in that state, in the switch below (this code would have to move down below the switch). This ensures the election owner is correctly removing expired elections.

Reply:
`confirmed` would also transition to `expired` it seems.